src/model/model.py

```diff
@@ -160,6 +160,105 @@ def apply_operation(self, arr, elements=[], operation='add'):
+
+        Parameters
+        ----------
+        elements: dict
+            this dictionary contains data (image or float/int...)
+        formula: str
+
+        Return
+        ------
+        result: np.ndarray or int/float
+            the result type depends on the formula.
+
+        Example
+        -------
+        formula = "( 3 x [key_1] + 4 x [key_2] ) ÷ [key_3]"
+        elements = {'key_1': np.array([...]),   # image
+                    'key_2': np.array([...]),   # image
+                    'key_3': 2}                 # integer
+
+        This will recursively apply (call self.apply_operation method):
+            - M1 = multiplication of key_1 image and 3
+            - M2 = multiplication of key_2 image and 4
+            - A1 = addition of M1 and M2
+            - result = division of A1 and key_3 value
+        """
+        # operations ordered on priorities
+        operations = ['multiply', 'divide', 'add', 'subtract']
+        signs = ['x', '÷', '+', '-']
+
+        # initialize formula
+        if initialize:
+            formula = formula.replace(',', '.')
+            # add space
+            formula = list(formula)
+            for i, f in enumerate(formula):
+                if f in signs + ['(', ')']:
+                    formula[i] = " {} ".format(f)
+                elif f == '[':
+                    formula[i] = " " + f
+                elif f == ']':
+                    formula[i] = f + " "
+            formula = "".join(formula)
+
+            # convert formula to list
+            if isinstance(formula, str):
+                formula = formula.split(' ')
+            while '' in formula:
+                formula.remove('')
+
+        # apply pear operation (with one operator and two elements)
+        # example [Load_1] x 5
+        if len(formula) == 3:
+            op_sign = formula.pop(1)
+            for i, f in enumerate(formula):
+                if f.startswith('['):
+                    formula[i] = elements[f[1:-1]]
+                else:
+                    formula[i] = eval(f)
+            for operation, sign in zip(operations, signs):
+                if op_sign == sign:
+                    res = self.apply_operation(formula[0], formula[1], operation)
+        else:
+            # recursively apply formula on group in parenthesis
+            if '(' in formula:
+                for i, f in enumerate(formula):
+                    if f == '(':
+                        start = i
+                    elif f == ')':
+                        end = i + 1
+                        break
+                sub_formula = formula[start+1:end-1]
+
+            # recursively apply formula on prioritary pear operations
+            else:
+                for sign in signs:
+                    if sign in formula:
+                        start = formula.index(sign) - 1
+                        end = formula.index(sign) + 2
+                        break
+                sub_formula = formula[start:end]
```

Comment:
If start and end are only used to filter formula, you can only use one `sign_idx`, and use `[sign_idx - 1: sign_idx + 2]`. But if you prefer explicitly have start and end, I'm okay with it 👍

Reply:
yes but I need them later to rewrite the formula..